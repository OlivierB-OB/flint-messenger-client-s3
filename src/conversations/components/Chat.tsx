import Box from '@material-ui/core/Box';
import Container from '@material-ui/core/Container';
import LinearProgress from '@material-ui/core/LinearProgress';
import React from 'react';
import { connect } from 'react-redux';
import { match as Match } from 'react-router-dom';
import { Action } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { ChatMessages } from './ChatMessages';
import { ChatInput } from './ChatInput';
import { IAppState } from '../../appReducer';
import { IConversation, IConversationsStatus } from '../types';
import { Alert } from '../../layout/components/Alert';
import { makeConversationSeen } from '../actions/makeConversationSeen';

export interface IChatDisplayProps {
  status: IConversationsStatus;
  conversationId?: string;
  conversation?: IConversation;
  conversationSeen: (id: string) => void;
}

export function ChatDisplay({ status, conversationId, conversation, conversationSeen }: IChatDisplayProps) {
  if (!conversationId || !conversation) return null;
  const progress = status === 'sending' ? <LinearProgress /> : null;
  return (
    <Container style={{ height: '100%', padding: '1rem', boxSizing: 'border-box' }}>
      <Box style={{ height: '5%' }}>
        <Alert status={status} />
        {progress}
      </Box>
      <Box style={{ height: '70%', overflow: 'auto' }}>
        <ChatMessages
          conversationId={conversationId}
          messages={conversation.messages}
          conversationSeen={conversationSeen}
        />
      </Box>
      <Box style={{ height: '20%' }}>
        <ChatInput conversationId={conversationId} />
      </Box>
    </Container>
  );
}

export interface IChatProps {
  match: Match<{ conversationId: string }>;
}

const mapStateToProps = ({ conversations }: IAppState, { match }: IChatProps) => ({
  status: conversations.status,
  conversationId: match.params?.conversationId,
  conversation: conversations.conversations.find(({ _id }) => {
    return _id === match.params?.conversationId;
  }),
});

const mapDispatchToProps = (dispatch: ThunkDispatch<IAppState, void, Action>) => ({
  conversationSeen: (id: string) => void dispatch(makeConversationSeen(id)),
});

export const Chat = connect(mapStateToProps, mapDispatchToProps)(ChatDisplay);
