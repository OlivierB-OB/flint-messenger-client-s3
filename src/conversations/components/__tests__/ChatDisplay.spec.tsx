import React from 'react';
import { shallow } from 'enzyme';
import { ChatDisplay } from '../Chat';

describe('ChatDisplay', () => {
  it('should display the chat', async () => {
    const props: any = {
      status: 'conversation status',
      conversationId: 'conversation id',
      conversation: { messages: 'conversation messages' },
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should display a progress bar while sending', async () => {
    const props: any = {
      status: 'sending',
      conversationId: 'conversation id',
      conversation: { messages: 'conversation messages' },
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should return null if conversationId has not been provided', async () => {
    const props: any = {
      status: 'conversation status',
      conversation: { messages: 'conversation messages' },
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });

  it('should return null if conversation has not been provided', async () => {
    const props: any = {
      status: 'conversation status',
      conversationId: 'conversation id',
      conversationSeen: jest.fn(),
    };
    const component = shallow(<ChatDisplay {...props} />);
    expect(component).toMatchSnapshot();
  });
});
