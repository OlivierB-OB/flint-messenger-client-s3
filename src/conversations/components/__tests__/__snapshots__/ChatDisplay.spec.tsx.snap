// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatDisplay should display a progress bar while sending 1`] = `
<WithStyles(ForwardRef(Container))
  style={
    Object {
      "boxSizing": "border-box",
      "height": "100%",
      "padding": "1rem",
    }
  }
>
  <Styled(MuiBox)
    style={
      Object {
        "height": "5%",
      }
    }
  >
    <Alert
      status="sending"
    />
    <WithStyles(ForwardRef(LinearProgress)) />
  </Styled(MuiBox)>
  <Styled(MuiBox)
    style={
      Object {
        "height": "70%",
        "overflow": "auto",
      }
    }
  >
    <ChatMessages
      conversationId="conversation id"
      conversationSeen={[MockFunction]}
      messages="conversation messages"
    />
  </Styled(MuiBox)>
  <Styled(MuiBox)
    style={
      Object {
        "height": "20%",
      }
    }
  >
    <Connect(ChatInputDisplay)
      conversationId="conversation id"
    />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Container))>
`;

exports[`ChatDisplay should display the chat 1`] = `
<WithStyles(ForwardRef(Container))
  style={
    Object {
      "boxSizing": "border-box",
      "height": "100%",
      "padding": "1rem",
    }
  }
>
  <Styled(MuiBox)
    style={
      Object {
        "height": "5%",
      }
    }
  >
    <Alert
      status="conversation status"
    />
  </Styled(MuiBox)>
  <Styled(MuiBox)
    style={
      Object {
        "height": "70%",
        "overflow": "auto",
      }
    }
  >
    <ChatMessages
      conversationId="conversation id"
      conversationSeen={[MockFunction]}
      messages="conversation messages"
    />
  </Styled(MuiBox)>
  <Styled(MuiBox)
    style={
      Object {
        "height": "20%",
      }
    }
  >
    <Connect(ChatInputDisplay)
      conversationId="conversation id"
    />
  </Styled(MuiBox)>
</WithStyles(ForwardRef(Container))>
`;

exports[`ChatDisplay should return null if conversation has not been provided 1`] = `""`;

exports[`ChatDisplay should return null if conversationId has not been provided 1`] = `""`;
